### ISSUE TITLE
Variable type is ignored with an array member as initial value

### ISSUE THREAD

**Issue by c0rrup7eq** [godotengine/godot#27361]
**Godot version:**
v3.1.stable.official

**OS/device including version:**
Windows 10 Home

**Issue description:**
When assigning an array member to a variable as an initial value, the variable changes its type and value disregarding any type you declare statically (though I haven't tested it with all types). In the editor such line is also marked as 'safe'.

But when you try to assign a new value to this variable, a wrong assignment error appears, so it somehow has 2 types at the same time.

**Steps to reproduce:**
1. Add `var a : Node = ["Word"][0]` to a script
2. The `a` will be of `String` type with a value of `Word`
3. Add `a = [1][0]` after the first line
4. This should create an error "Trying to assign value of type `int` to a variable of type `Node`"

**Minimal reproduction project:**
[Typed Test.zip](https://github.com/godotengine/godot/files/2999798/Typed.Test.zip)

**Comment by Zylann** [comment]
`["Word"]` and `[1]` are Arrays, which store variables of any type. That means the static checks cannot be sure you are getting a value of the correct type out of them, even though those arrays are declared using constants. None of those lines should be marked as safe, unless you add a cast to `Node` or `int`, or the static checks need ability to recognize composite constants like this.

**Comment by c0rrup7eq** [comment]
If casting is used on the first line then it is marked as unsafe. But I have a question on how the initial value is assigned, because it still has to get the value out of the array. Maybe the order is wrong? My guess is that when the program is run the type check happens and fails before the value is extracted from the array and it should be vice versa.

**Comment by Zaerei** [comment]
I understand why this is done, but personally I think there should at least be a build toggle (basically a compiler flag) in the settings of some sort that automatically expands

```
x: SomeType = someVariant
```

to essentially

```
x: SomeType = someVariant
assert(x is SomeType)
```

so we can at least get runtime checking instead of silent failure.

This can be expensive hence making it a toggle, but I find myself doing this defensively just in case assigning from an Array, Dictionary, etc goes wrong. I've had type bugs due to mishandled assignments in other dynamic languages before, they're never fun tracking down.

**Comment by vnen** [comment]
This was solved by #34916.
